pkg/nfs/controllerserver.go

```diff
@@ -10,14 +15,134 @@ import (
 
 type ControllerServer struct {
 	Driver *nfsDriver
+	// Working directory for the provisioner to temporarily mount nfs shares at
+	workingMountDir string
+}
+
+// nfsVolume is an internal representation of a volume
+// created by the provisioner.
+type nfsVolume struct {
+	// Volume id
+	id string
+	// Address of the NFS server.
+	// Matches paramServer.
+	server string
+	// Base directory of the NFS server to create volumes under
+	// Matches paramShare.
+	baseDir string
+	// Subdirectory of the NFS server to create volumes under
+	subDir string
+	// size of volume
+	size int64
 }
 
+// Ordering of elements in the CSI volume id.
+// ID is of the form {server}/{baseDir}/{subDir}.
+// TODO: This volume id format limits baseDir and
+// subDir to only be one directory deep.
+// Adding a new element should always go at the end
+// before totalIDElements
+const (
+	idServer = iota
+	idBaseDir
+	idSubDir
+	totalIDElements // Always last
+)
+
 func (cs *ControllerServer) CreateVolume(ctx context.Context, req *csi.CreateVolumeRequest) (*csi.CreateVolumeResponse, error) {
-	return nil, status.Error(codes.Unimplemented, "")
+	// Validate arguments
+	name := req.GetName()
+	if len(name) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "CreateVolume name must be provided")
+	}
+
+	if err := cs.validateVolumeCapabilities(req.GetVolumeCapabilities()); err != nil {
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+
+	reqCapacity := req.GetCapacityRange().GetRequiredBytes()
+	nfsVol, err := cs.newNFSVolume(name, reqCapacity, req.GetParameters())
+	if err != nil {
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+
+	// Mount nfs base share so we can create a subdirectory
+	if err = cs.internalMount(ctx, nfsVol); err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to mount nfs server: %v", err.Error())
+	}
+	defer func() {
+		if err = cs.internalUnmount(ctx, nfsVol); err != nil {
+			glog.Warningf("failed to unmount nfs server: %v", err.Error())
+		}
+	}()
+
+	// Create subdirectory under base-dir
+	// TODO: revisit permissions
+	internalVolumePath := cs.getInternalVolumePath(nfsVol)
+
+	if volumePathExists(internalVolumePath) {
+		if err := validCapacity(reqCapacity, internalVolumePath); err != nil {
+			return nil, err
+		}
+		return &csi.CreateVolumeResponse{}, nil
+	}
+
+	if err = os.Mkdir(internalVolumePath, 0777); err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to make subdirectory: %v", err.Error())
+	}
+	// Remove capacity setting when provisioner 1.4.0 is available with fix for
+	// https://github.com/kubernetes-csi/external-provisioner/pull/271
+	return &csi.CreateVolumeResponse{Volume: cs.nfsVolToCSI(nfsVol, reqCapacity)}, nil
```

Comment:
@msau42 I was also a little confused about this part. Should the response not contain information about capacity?

Reply:
this was a workaround until the issue in external-provisioner was fixed. Now that it is fixed, I think we can return 0 capacity.